### PR TITLE
WIP: KFService v1alpha2 type for PrePostProcess, Predictor and Explainer

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:
Add KFService v1alpha2 type to support static graph for pre/post processing, predictor and explainer

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubeflow/kfserving/issues/265

**Special notes for your reviewer**:

- Add `DeploymentSpec` to include resources and min/max replicas
- Move v1alpha1 `ModelSpec` down to one level under `Predictor`
- Add `ServiceSpec` to include `PrepostProcess`, `Predictor`, `Explain` services

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/283)

### CODE REVIEW

File: pkg/apis/serving/v1alpha1/kfservice_types.go [hunk 1]
```diff
@@ -57,11 +57,11 @@ type ExplainSpec struct {
 type AlibiExplainerType string
 
 const (
```
Comment (rakelkar) [inline pkg/apis/serving/v1alpha1/kfservice_types.go]: move to `v1alpha2` folder?

File: pkg/apis/serving/v1alpha2/framework.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"fmt"
+
+	"github.com/kubeflow/kfserving/pkg/constants"
+	v1 "k8s.io/api/core/v1"
+	resource "k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog"
+)
+
+type FrameworkHandler interface {
```
Comment (rakelkar) [inline pkg/apis/serving/v1alpha2/framework.go]: seems annoying to have to duplicate all this code, but cant think of a way to avoid it...

Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha2/framework.go]: Duck typing? :)

File: pkg/apis/serving/v1alpha2/kfservice_types.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+	Custom     *CustomSpec     `json:"custom,omitempty"`
+	Tensorflow *TensorflowSpec `json:"tensorflow,omitempty"`
+	TensorRT   *TensorRTSpec   `json:"tensorrt,omitempty"`
+	XGBoost    *XGBoostSpec    `json:"xgboost,omitempty"`
+	SKLearn    *SKLearnSpec    `json:"sklearn,omitempty"`
+	PyTorch    *PyTorchSpec    `json:"pytorch,omitempty"`
+}
+
+type PrepostprocessSpec struct {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/kfservice_types.go]: I feel a little bit that Preprocessor and PostProcessor and a little unwieldy as terms.

I wonder if there's another way we could think of them. I kind of like Seldon's "Transformer" concept. Perhaps we could take a list of transformers and part of the transformer spec could be `type={"before", "after", "both" (default)}. We'd need some semantic for what order they're executed in, but I think FIFO is probably clean enough.

What do you think?

File: pkg/apis/serving/v1alpha2/kfservice_types.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	Predict ModelSpec `json:"predict"`
+	// Optional Explain specification to add a model explainer next to the chosen predictor.
+	Explain *ExplainSpec `json:"explain,omitempty"`
+	// Default deployment specification
+	Deployment *DeploymentSpec `json:"deployment,omitempty"`
+	// Service account name
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+type DeploymentSpec struct {
```
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/kfservice_types.go]: Cool. I think this is a great abstraction. Should it also include ServiceAccountName?

File: pkg/apis/serving/v1alpha2/kfservice_types.go (hunk 3 above)
Comment (ellis-bigelow) [inline pkg/apis/serving/v1alpha2/kfservice_types.go]: I'm a little confused by coupling pre and post. Does that mean that it's executed twice? What if no post-processor is necessary?

Comment (rakelkar) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: +1 to transformer..

Comment (cliveseldon) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: Also thinking a bit ahead and a graph sitting above kfserving each node if really doing a "transform". Might be great to allow "transform" to do whole sequence of pre->predict->post. That could be part of data plane discussions.

File: pkg/apis/serving/v1alpha2/kfservice_types.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	Default ServicesSpec `json:"default"`
+	// Canary defines an alternate KFService configuration to route a percentage of traffic.
+	Canary *ServicesSpec `json:"canary,omitempty"`
+	// Traffic split percentage between default and canary KFService
+	CanaryTrafficPercent int `json:"canaryTrafficPercent,omitempty"`
+}
+
+type ServicesSpec struct {
+	// Optional prepost processing specification
+	Prepostprocess *PrepostprocessSpec `json:"prepostprocess,omitempty"`
```
Comment (animeshsingh) [inline pkg/apis/serving/v1alpha2/kfservice_types.go]: Should we capitalize - something like this ? PrePostProcessSpec

File: pkg/apis/serving/v1alpha2/kfservice_types.go (hunk 3 above)
Comment (animeshsingh) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: PreProcessor and PostProcessor are well accepted in industry - and most of the users would understand what they mean in context of models.

Comment (yuzisun) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: This is from efficiency perspective, most of time post processing is really light(e.g map resulting tensor to label) and not sure if worth to separate into different pod or different container. User can implement Preprocess(required) and Postprocess function(can be noop), the service wires up calling predictor after preprocess and then postprocess.

Comment (yuzisun) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: I am reading from here https://www.tensorflow.org/tfx/transform/get_started, we can name it transformer on the spec and user implements preprocess_fn and postprocess_fn according to our kfserver interface.

Comment (ellis-bigelow) [reply pkg/apis/serving/v1alpha2/kfservice_types.go]: I like this a lot!

File: pkg/apis/serving/v1alpha2/kfservice_status.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	}
+}
+
+// PropagateCanaryConfigurationStatus propagates the canary Configuration status and applies its values
+// to the Service status.
+func (ss *KFServiceStatus) PropagateCanaryConfigurationStatus(canaryConfigurationStatus *knservingv1alpha1.ConfigurationStatus) {
+	// reset status if canaryConfigurationStatus is nil
+	if canaryConfigurationStatus == nil {
+		ss.Canary = StatusConfigurationSpec{}
+		conditionSet.Manage(ss).MarkUnknown(CanaryPredictorReady, "CanarySpecUnavailable", "Canary spec unavailable")
```
Comment (rakelkar) [inline pkg/apis/serving/v1alpha2/kfservice_status.go]: this copy of the code is so sad... it misses for instance the fix just merged with #285 
can we please have a strategy that does not require duplication of all the code?
or - just bite the bullet - delete the old code and move to the new path quickly to minimize divergence?
@johnugeorge  @ellis-bigelow

File: pkg/apis/serving/v1alpha2/kfservice_types.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+	"github.com/knative/pkg/apis"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// KFServiceSpec defines the desired state of KFService
+type KFServiceSpec struct {
+	// Default defines the configuration of KFService
+	Default ServicesSpec `json:"default"`
```
Comment (rakelkar) [inline pkg/apis/serving/v1alpha2/kfservice_types.go]: ServiceSpec? after all its just a KFService with multiple parts (still one thing)

File: pkg/apis/serving/v1alpha2/kfservice_status.go (hunk 6 above)
Comment (yuzisun) [reply pkg/apis/serving/v1alpha2/kfservice_status.go]: Let's close this PR and start with a new PR just renaming v1alpha1 to v1alpha2, then add v1alpha2 changes on top of that, what do you guys think ?

**Comment by ellis-bigelow** [comment]
Sgtm

File: pkg/apis/serving/v1alpha2/kfservice_status.go (hunk 6 above)
Comment (rakelkar) [reply pkg/apis/serving/v1alpha2/kfservice_status.go]: Makes sense!

**Comment by yuzisun** [comment]
closing this for https://github.com/kubeflow/kfserving/pull/290

**Pull request closed without merging**